### ISSUE TITLE
MIOpen using clang-14 fails 4 tests: "invalid HSA metadata", 1 

### ISSUE THREAD

**Issue by littlewu2508** [ROCm/MIOpen#1731]
I'm packeging MIOpen 5.1.3 for gentoo, and noticing there are test failures.

All errors are "invalid HSA metadata" when compiling assemblies:

```
AMD_COMGR_SAVE_TEMPS=1 AMD_COMGR_REDIRECT_LOGS=stdout AMD_COMGR_EMIT_VERBOSE /test_immed_conv2d --float --cmode conv --pmode default --group-count 1 --enable-fdb 0 --batch_size 8 --input_channels 16 --output_channels 64 --spatial_dim_elements 28 28 --filter_dims 5 5 --pads_strides_dilations 1 1 1 1 1 1 --trans_output_pads 0 0 --in_layout NCHW --fil_layout NCHW --out_layout NCHW

...

COMGR::executeInProcessDriver argv: clang "-cc1as" "-triple" "amdgcn-amd-amdhsa" "-filetype" "obj" "-main-file-name" "Conv_Winograd_v21_1_3_fp32_stride1.s" "-target-cpu" "gfx1031" "-target-feature" "+wavefrontsize64" "-target-feature" "+cumode" "-I" "/tmp/comgr-1d8e33/include" "-fdebug-compilation-dir=/home/selene" "-dwarf-debug-producer" "clang version 14.0.6" "-I" "/tmp/comgr-1d8e33/include" "-dwarf-version=5" "-mrelocation-model" "pic" "-defsym" "ROCM_METADATA_VERSION=5" "--mrelax-relocations" "-o" "/tmp/comgr-1d8e33/output/Conv_Winograd_v21_1_3_fp32_stride1.s.o" "/tmp/comgr-1d8e33/input/Conv_Winograd_v21_1_3_fp32_stride1.s"

<instantiation>:67:21: error: invalid HSA metadata
.end_amdgpu_metadata
    |   |   |   |   ^
<instantiation>:1:1: note: while in macro instantiation
METADATA %total_sgpr_count, %.amdgcn.next_free_vgpr, %workgroup_size_x, miopenSp3AsmConv_v21_1_3_gfx10_fp32_stride1
^
<instantiation>:93:1: note: while in macro instantiation
METADATA_WRAPPER total_sgpr_count,.amdgcn.next_free_vgpr,workgroup_size_x, <miopenSp3AsmConv_v21_1_3_gfx10_fp32_stride1>
^
<instantiation>:1:1: note: while in macro instantiation
EPILOG_KERNEL_DESCRIPTOR miopenSp3AsmConv_v21_1_3_gfx10_fp32_stride1
^
<instantiation>:1:1: note: while in macro instantiation
EPILOG_KERNEL_DESCRIPTOR_WRAPPER %.amdgcn.gfx_generation_number, fp32_stride1
^
/tmp/comgr-1d8e33/input/Conv_Winograd_v21_1_3_fp32_stride1.s:9648:1: note: while in macro instantiation
KERNEL_EPILOG fp32_stride1
^
    ReturnStatus: AMD_COMGR_STATUS_ERROR

```

I wonder what difference between llvm/clang from AMD and upstream llvm/clang is causing this?

The complete test log is: 
[miopen-vanilla-clang-test.log](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9471708/miopen-vanilla-clang-test.log)

**Comment by junliume** [comment]
@atamazov  could you help to take a look?

**Comment by littlewu2508** [comment]
So I find the metadata is from src/kernels/Conv_Winograd_v21_1_3_metadata.inc

Can confirm that clang from ROCm-5.1.3 release can built the assembly without issue, while clang-14.0.6 thinks the metadata is invalid

**Comment by littlewu2508** [comment]
I guess this is because https://reviews.llvm.org/D118272 added new metadata requirements:

https://github.com/llvm/llvm-project/blob/llvmorg-14.0.6/llvm/lib/BinaryFormat/AMDGPUMetadataVerifier.cpp requires more metadata entries

**Comment by littlewu2508** [comment]
From gdb tracing, I find the metadata verification fails at https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/lib/BinaryFormat/AMDGPUMetadataVerifier.cpp#L99

**Comment by littlewu2508** [comment]
I comment out the metadata wrapper in src/kernels/Conv_Winograd_v21_1_3_metadata.inc, which makes one test passed. A temp workaround maybe.

**Comment by atamazov** [comment]
@junliume It looks like the reason of the issue is that https://reviews.llvm.org/D118272 has broken support for V3 metadata.  Some V5-specific checks were **_unconditionally_** added to MetadataVerifier, see https://github.com/llvm/llvm-project/commit/1194b9cdda30df3b76f8dc932d438dc1b3f3ed6c#r82920767

@littlewu2508 Thanks for triaging the issue!

**Comment by littlewu2508** [comment]
> @junliume It looks like the reason of the issue is that https://reviews.llvm.org/D118272 has broken support for V3 metadata. Some V5-specific checks were **_unconditionally_** added to MetadataVerifier, see [llvm/llvm-project@1194b9c#r82920767](https://github.com/llvm/llvm-project/commit/1194b9cdda30df3b76f8dc932d438dc1b3f3ed6c#r82920767)

OK. I tried to delete these lines and reinstall llvm, but turns out still getting invalid HSA metadata error. Actually,  

> From gdb tracing, I find the metadata verification fails at https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/lib/BinaryFormat/AMDGPUMetadataVerifier.cpp#L99

Indicates that the verification already fails at ".name" entry.

Further tracing, I finally get the invalid point:https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/lib/BinaryFormat/AMDGPUMetadataVerifier.cpp#L34

The entry is a llvm::msgpack::Type::Boolean but the required type is llvm::msgpack::Type::String, so not passing the (non-strict) check. But I don't know why it is a boolean.

**Comment by littlewu2508** [comment]
I figured it out:

Conv_Winograd_v21_1_3_metadata.inc:54
```assembly
 - { .size: 4, .offset:   0, .value_kind: by_value, .value_type: i32, .name: N }
```

This `N` is recognized as "false", a boolean type. This explains why I get llvm::msgpack::Type::Boolean, not llvm::msgpack::Type::String.

I chaned it into other characters and the error disappears.

**Comment by littlewu2508** [comment]
I tried clang-12, clang-13 and clang-15 and they both behave the same, so I guess there is some difference between the upstream llvm and https://github.com/RadeonOpenCompute/llvm-project/

**Comment by XCRailFans** [comment]
> ```assembly
>  - { .size: 4, .offset:   0, .value_kind: by_value, .value_type: i32, .name: N }
> ```

This is YAML syntax. And famously in YAML, unquoted literals tend to get recognized as random things instead of strings. 🤷

**Comment by littlewu2508** [comment]
@atamazov 
I changed the name "N" to another character, for example X, and test_conf2d passed successfully.

Looking at the source code of https://github.com/ROCmSoftwarePlatform/MIOpen/blob/f4538d74520ee80aa2f2cae792d743e7257d40f5/src/solver/conv_winoRxS.cpp#L864 I believe the `.name: N` entry is only for matching the cpp code, and does not harm if change the name. I that correct?

**Comment by atamazov** [comment]
@littlewu2508 This is amazing finding, thanks again. Changing the name should not affect anything, only types of parameters should be checked at runtime.

Can you please provide me with reproduce instructions? I need to look at this. I would like to make the fix as sustainable as possible to future compiler changes.

**Comment by atamazov** [comment]
@littlewu2508 I suspect this problem is very important for you, but could you please clearly indicate it? Thank you!

@junliume Can you please set the urgency label as appropriate? Thanks!

**Comment by littlewu2508** [comment]
> Can you please provide me with reproduce instructions? I need to look at this. I would like to make the fix as sustainable as possible to future compiler changes.

For minimal bug reproduction and simple workaround by changing `N` into other names:

I extracted the assembly via `AMD_COMGR_SAVE_TEMPS=1 AMD_COMGR_REDIRECT_LOGS=stdout AMD_COMGR_EMIT_VERBOSE_LOGS=1 ./test_conv2d  --float --all --limit 2 --tolerance 130`, and the gziped code is:

[Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s.gz](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9493243/Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s.gz)

Using vanilla clang-14 (not the clang from AMD), compile the assembly using following command:

```
clang -cc1as -triple amdgcn-amd-amdhsa -filetype obj -main-file-name Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s -target-cpu gfx1030 -target-feature +wavefrontsize64 -target-feature +cumode -mrelocation-model pic -defsym ROCM_METADATA_VERSION=5 --mrelax-relocations -o Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s.o Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s
```

The error of "invalid HSA metadata" occurs. Edit the Conv_Winograd_v21_1_3_f3x2_fp32_stride1_group.s line 79, change `.name: N` to `.name: X`, and the compilation is successful.

**Comment by atamazov** [comment]
@littlewu2508 Thank you. Will update this ticket when the full-blown fix ready.

**Comment by atamazov** [comment]
@littlewu2508 Please review #1816, thanks.

**Comment by littlewu2508** [comment]
Sorry I forgot to mention, that this issue also exists in other kernels -- the lower case "n" and "y" are affected.

For example, in [src/kernels/dynamic_igemm/igemm_v4r1_dynamic.s](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/kernels/dynamic_igemm/igemm_v4r1_dynamic.s), those `.name: n` and `.name: y` has to be replaced, too. That's a huge modification, affecting thousands of kernel files.

So https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1816 solves `test_immed_conv2d` when using upstream clang, but there are still failures.

Currently in Gentoo `sci-libs/miopen` package, I use
```bash
find src/kernels -name "*.s" -exec \
      sed -e "s/.name: n /.name: x /g" -e "s/.name: y /.name: z /g" \
             -e "s/.name: y,/.name: z,/g" -i {} \;
```
to do the replacement (I use meaningless `x` and `z`). Then all tests can pass

**Comment by junliume** [comment]
@atamazov it seems that #1816 does not fix them all, and there will be more patches incoming?

**Comment by littlewu2508** [comment]
Because https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1816 does not fix all the cases, can this issue be reopened?

**Comment by atamazov** [comment]
@littlewu2508 @junliume Please reopen (I can't)

**Comment by atamazov** [comment]
@littlewu2508 Thanks for triaging and for the script. I am going to append underscore to n and y. Please expect another PR very soon.

**Comment by atamazov** [comment]
@littlewu2508 #1830 created 😎

**Comment by ppanchad-amd** [comment]
@atamazov Is this ticket still relevant? Thanks!

**Comment by atamazov** [comment]
@ppanchad-amd Fixed by #1830, please close.
